### PR TITLE
Add max background jobs to configuration

### LINKED ISSUE #1292: Enable max_background_jobs
### Discussed in https://github.com/apache/incubator-kvrocks/discussions/1284

<div type='discussions-op-text'>

<sup>Originally posted by **xiaobiaozhao** February 27, 2023</sup>
Use `max_background_jobs` instead of `max_background_compactions` and `max_background_flushes`

https://github.com/facebook/rocksdb/blob/main/include/rocksdb/options.h#L699</div>

### PR DESCRIPTION
**Pull request by bruceqiwang**
This closes #1292

### CODE REVIEW

**Comment by torwig** [comment]
@bruceqiwang Shouldn't we set the deprecated options here: https://github.com/bruceqiwang/incubator-kvrocks/blob/max_background_jobs_new/src/storage/storage.cc#L130 ? And the following 131 line as well?

**Review by PragmaTwice — COMMENTED** [review]
Could you make an explanation about why you create a new PR for this patch?

Why do not continue the development in #1328?

**Comment by bruceqiwang** [comment]
i want to rewrite code and checker failed. @PragmaTwice

**Comment by bruceqiwang** [comment]
@torwig when the config is deprecated, all the invalid code need to delete? I will check and delete no used code

**Comment by torwig** [comment]
@bruceqiwang Yes, you are right, there is no reason to delete that code for now.

**Review by torwig — DISMISSED** [review]
LGTM

File: kvrocks.conf [hunk 1]
```diff
@@ -587,15 +587,19 @@ rocksdb.target_file_size_base 128
 # allowed.
 rocksdb.max_write_buffer_number 4
 
-# Maximum number of concurrent background compaction jobs, submitted to
+# Maximum number of concurrent background jobs (compactions and flushes).
+# RocksDB automatically decides this based on the  value of max_background_jobs. For backwards compatibility we will set `max_background_jobs = max_background_compactions + max_background_flushes` in the case where user sets at least one of `max_background_compactions` or `max_background_flushes` (we replace -1 by 1 in case one option is unset).
```
Comment (PragmaTwice) [inline kvrocks.conf]: Seems it is better to split this comment to multiple lines like all other comments in this file.

File: kvrocks.conf [hunk 2]
```diff
@@ -587,15 +587,19 @@ rocksdb.target_file_size_base 128
...
+ rocksdb.max_background_jobs 4
+
+# DEPRECATED : Maximum number of concurrent background compaction jobs, submitted to
```
Comment (PragmaTwice) [inline kvrocks.conf]: You need to specify a reason for DEPRECATED.

e.g. `# DEPRECATED: the reason, and what users can do`

File: kvrocks.conf [hunk 3]
```diff
@@ -587,15 +587,23 @@ rocksdb.target_file_size_base 128
 # allowed.
 rocksdb.max_write_buffer_number 4
 
+# Maximum number of concurrent background jobs (compactions and flushes).
+# RocksDB automatically decides this based on the  value of max_background_jobs.# For backwards compatibility we will set `max_background_jobs =
```
Comment (PragmaTwice) [inline kvrocks.conf]: Why is there a `#` in the middle of this line?

File: kvrocks.conf [hunk 4]
```diff
@@ -587,15 +587,23 @@ rocksdb.target_file_size_base 128
 # allowed.
 rocksdb.max_write_buffer_number 4
 
+# Maximum number of concurrent background jobs (compactions and flushes).
+# RocksDB automatically decides this based on the  value of max_background_jobs.# For backwards compatibility we will set `max_background_jobs = 
+# max_background_compactions + max_background_flushes` in the case where user
+# sets at least one of `max_background_compactions` or `max_background_flushes` # (we replace -1 by 1 in case one option is unset).
+ rocksdb.max_background_jobs 4
+
+# DEPRECATED : RocksDB automatically decides this based on the value of max_background_jobs
```
Comment (PragmaTwice) [inline kvrocks.conf]:
```suggestion
# DEPRECATED: it is automatically decided based on the value of rocksdb.max_background_jobs
```

File: kvrocks.conf [hunk 5]
```diff
@@ -587,15 +587,23 @@ rocksdb.target_file_size_base 128
...
 # Maximum number of concurrent background compaction jobs, submitted to
 # the default LOW priority thread pool.
-rocksdb.max_background_compactions 4
+rocksdb.max_background_compactions -1
 
+# DEPRECATED : RocksDB automatically decides this based on the value of max_background_jobs
```
Comment (PragmaTwice) [inline kvrocks.conf]: ditto

File: kvrocks.conf [hunk 6]
```diff
@@ -588,17 +588,19 @@ rocksdb.target_file_size_base 128
 rocksdb.max_write_buffer_number 4
 
 # Maximum number of concurrent background jobs (compactions and flushes).
-# RocksDB automatically decides this based on the  value of max_background_jobs.# For backwards compatibility we will set `max_background_jobs = 
-# max_background_compactions + max_background_flushes` in the case where user
-# sets at least one of `max_background_compactions` or `max_background_flushes` # (we replace -1 by 1 in case one option is unset).
+# RocksDB automatically decides this based on the  value of max_background_jobs.
```
Comment (PragmaTwice) [inline kvrocks.conf]: This line does not make sense. `rocksdb.max_background_jobs` cannot be decided by its own.

Please be careful.

File: kvrocks.conf [hunk 7]
```diff
@@ -587,15 +587,24 @@ rocksdb.target_file_size_base 128
 # allowed.
 rocksdb.max_write_buffer_number 4
 
+# Maximum number of concurrent background jobs (compactions and flushes).
+# For backwards compatibility we will set `max_background_jobs =
+# max_background_compactions  + max_background_flushes` in the case where user
```
Comment (PragmaTwice) [inline kvrocks.conf]:
```suggestion
# max_background_compactions + max_background_flushes` in the case where user
```

File: kvrocks.conf [hunk 8]
```diff
@@ -587,15 +587,24 @@ rocksdb.target_file_size_base 128
...
+# (we replace -1 by 1 in case one option is unset).
+rocksdb.max_background_jobs 4
+
+# DEPRECATED : RocksDB automatically decides this based on the value of max_background_jobs.
 # Maximum number of concurrent background compaction jobs, submitted to
 # the default LOW priority thread pool.
-rocksdb.max_background_compactions 4
+rocksdb.max_background_compactions -1
 
+# DEPRECATED : RocksDB automatically decides this based on the value of max_background_jobs.
```
Comment (PragmaTwice) [inline kvrocks.conf]:
```suggestion
# DEPRECATED: it is automatically decided based on the value of rocksdb.max_background_jobs
```

File: kvrocks.conf [hunk 9]
```diff
@@ -587,15 +587,24 @@ rocksdb.target_file_size_base 128
...
 rocksdb.max_write_buffer_number 4
 
+# Maximum number of concurrent background jobs (compactions and flushes).
+# For backwards compatibility we will set `max_background_jobs =
+# max_background_compactions + max_background_flushes` in the case where user
+# sets at least one of `max_background_compactions` or `max_background_flushes`
+# (we replace -1 by 1 in case one option is unset).
+rocksdb.max_background_jobs 4
+
+# DEPRECATED : RocksDB automatically decides this based on the value of max_background_jobs.
```
Comment (PragmaTwice) [inline kvrocks.conf]:
```suggestion
# DEPRECATED: it is automatically decided based on the value of rocksdb.max_background_jobs
```

**Review by git-hulk — APPROVED** [review]

**Review by PragmaTwice — APPROVED** [review]

**Review by torwig — APPROVED** [review]

**Comment by PragmaTwice** [comment]
Thanks for your contribution. Merging...

**Pull request merged**